compiler/rustc_mir_build/src/build/coverageinfo.rs

```diff
@@ -79,21 +84,92 @@ impl BranchInfoBuilder {
         }
     }
 
+    fn visit_coverage_branch_condition<'tcx>(
+        &mut self,
+        cfg: &mut CFG<'tcx>,
+        thir: &Thir<'tcx>,
+        scope: SourceScope,
+        mut expr_id: ExprId,
+        mut then_block: BasicBlock,
+        mut else_block: BasicBlock,
+    ) {
+        // If this condition expression is nested within one or more `!` expressions,
+        // replace it with the enclosing `!` collected by `visit_unary_not`.
+        if let Some(&NotInfo { enclosing_not, is_flipped }) = self.nots.get(&expr_id) {
+            expr_id = enclosing_not;
+            if is_flipped {
+                std::mem::swap(&mut then_block, &mut else_block);
+            }
+        }
+
+        let span = thir[expr_id].span;
+        let source_info = SourceInfo { span, scope };
+        let true_marker = self.inject_block_marker(cfg, source_info, then_block);
+        let false_marker = self.inject_block_marker(cfg, source_info, else_block);
+
+        self.branch_arm_lists.push(vec![
+            BranchArm { span, marker: true_marker },
+            BranchArm { span, marker: false_marker },
+        ]);
+    }
+
+    fn visit_coverage_match_expr(
+        &mut self,
+        cfg: &mut CFG<'_>,
+        scope: SourceScope,
+        arms: impl Iterator<Item = (Span, Option<BasicBlock>)>,
+    ) {
+        let Some(arms) = arms
+            .map(|(span, pre_binding_block)| Some((span, pre_binding_block?)))
+            .collect::<Option<Vec<_>>>()
+        else {
+            return;
+        };
```

Comment:
This is what causes us to currently not handle or-patterns, since we require every arm to have a `pre_binding_block`.

We can probably dig deeper into the pattern and find a suitable block, but I'll have to pay close attention to make sure that what I'm doing is correct.